### PR TITLE
Extend a softmax function and the loss layer to support replicated softmax units

### PR DESCRIPTION
**Pull request by rezoo**
This PR extends an existing Softmax function and a SoftmaxCrossEntropy function to support replicated softmax units.

Neural Network models that output multiple discrete values have been frequently used in many fields such as NLP, CV, and audio recognition.
While caffe already supports replicated softmax layer, the current implementation of chainer does not support the replicated softmax units (i.e., SoftmaxFunction only accepts a two dimensional input array).
In this PR, if we pass d (d>2) dimensional array to the SoftmaxFunction, it regards the remaining axis as the replicated units, and computes the replicated softmax.

Please note that currently I assume the second axis as a channel and the remaining axis as replicated units. The softmax layer in caffe already supports the ``axis`` parameter for indicating the channel, however, I do not implement it.

### CODE REVIEW

**Comment by beam2d** [comment]
Sorry for being late. I will review it.

File: chainer/functions/softmax.py [hunk 1]
```diff
@@ -50,46 +48,65 @@ def forward_cpu(self, x):
...
-                'float* y, const float* x, const float* maxes, int c',
-                'y[i] = __expf(x[i] - maxes[i / c])',
-                'softmax_exp')(y, x[0], maxes, c)
+                '''
+                   float* y, const float* x, const float* maxes,
+                   int n_channel, int n_unit
+                ''',
+                '''
+                   const int n = i / (n_channel * n_unit);
+                   const int m = (i % (n_channel * n_unit)) % n_unit;
```
Comment (beam2d) [inline chainer/functions/softmax.py]: ``m = i % n_unit`` is enough

File: chainer/functions/softmax.py [hunk 2]
```diff
@@ -50,46 +48,65 @@ def forward_cpu(self, x):
...
                    coeff[i] = 1 / sum;
-                ''', 'softmax_invrowsum')(coeff, y, c)
+                ''', 'softmax_invrowsum')(coeff, y, c, n_unit)
             cuda.elementwise(
-                'float* y, const float* coeff, int c', 'y[i] *= coeff[i / c]',
-                'softmax_rowmul')(y, coeff, c)
+                'float* y, const float* coeff, int n_channel, int n_unit',
+                '''
+                   const int n = i / (n_channel * n_unit);
+                   const int m = (i % (n_channel * n_unit)) % n_unit;
```
Comment (beam2d) [inline chainer/functions/softmax.py]: ditto

File: chainer/functions/softmax.py [hunk 3]
```diff
@@ -113,21 +131,31 @@ def backward_gpu(self, x, gy):
...
-                'float* gx, const float* y, const float* sum_ydy, int c',
-                'gx[i] -= y[i] * sum_ydy[i / c]',
-                'softmax_bwd_diff')(gx, self.y, sum_ydy, c)
+                '''
+                   float* gx, const float* y, const float* sum_ydy,
+                   int n_channel, int n_unit
+                ''',
+                '''
+                   const int n = i / (n_channel * n_unit);
+                   const int m = (i % (n_channel * n_unit)) % n_unit;
```
Comment (beam2d) [inline chainer/functions/softmax.py]: ditto

File: chainer/functions/softmax_cross_entropy.py [hunk 4]
```diff
@@ -38,49 +39,104 @@ def check_type_backward(self, in_types, out_types):
     def forward_cpu(self, inputs):
         x, t = inputs
         self.y, = softmax.Softmax().forward_cpu((x,))
-        p = self.y[six.moves.range(len(t)), t]
-        y = -numpy.log(p).sum(keepdims=True) / t.size
+        yd = self.y.transpose(
+            [0] + list(six.moves.range(2, self.y.ndim)) + [1])
+        yd = yd.reshape(numpy.prod(yd.shape[:-1]), -1)
```
Comment (beam2d) [inline chainer/functions/softmax_cross_entropy.py]: yd can be a bit simplified as follows:
```
yd = numpy.rollaxis(self.y, 1)
yd = yd.reshape(len(yd), -1).T
```

File: chainer/functions/softmax_cross_entropy.py [hunk 5]
```diff
@@ -38,49 +39,104 @@ def check_type_backward(self, in_types, out_types):
...
+        p = yd[six.moves.range(t.size), t.flat]
+        # deal with the case where the SoftmaxCrossEntropy is
+        # unpickled from the old version
+        if getattr(self, "normalize", True):
```
Comment (beam2d) [inline chainer/functions/softmax_cross_entropy.py]: Use single quotes for style consistency

File: chainer/functions/softmax_cross_entropy.py [hunk 6]
```diff
@@ -38,49 +39,104 @@ def check_type_backward(self, in_types, out_types):
...
         ret = cuda.reduce(
-            'int* t, float* y, int n_channel', '-log(y[i * n_channel + t[i]])',
+            'int* t, float* y, int n_channel, int n_unit',
+            '-log(y[n_unit * ((i / n_unit) * n_channel + t[i])'
+            '       + (i % n_unit)])',
             'a+b', '0', 'crossent_fwd', numpy.float32
-        )(t, self.y, self.y.shape[1])
-        ret /= t.size
+        )(t, self.y, self.y.shape[1], n_unit)
+        if getattr(self, "normalize", True):
```
Comment (beam2d) [inline chainer/functions/softmax_cross_entropy.py]: Use single quotes

File: chainer/functions/softmax_cross_entropy.py [hunk 7]
```diff
@@ -38,49 +39,104 @@ def check_type_backward(self, in_types, out_types):
...
+            # in the case where y.ndim is higher than 2,
+            # we think that a current implementation is inefficient
+            # because it yields two provisional arrays for indexing.
+            gx = self.y.copy().reshape(self.y.shape[0], self.y.shape[1], -1)
+            fst_index = numpy.arange(t.size) // n_unit
+            trd_index = numpy.arange(t.size) % n_unit
+            gx[fst_index, t.flat, trd_index] -= 1
+            gx = gx.reshape(self.y.shape)
+
+        if getattr(self, "normalize", True):
```
Comment (beam2d) [inline chainer/functions/softmax_cross_entropy.py]: ditto

File: chainer/functions/softmax_cross_entropy.py [hunk 8]
```diff
@@ -38,49 +39,104 @@ def check_type_backward(self, in_types, out_types):
...
+            count = t.shape[0]
+        gx *= gloss / count
         return gx, None
 
     def backward_gpu(self, inputs, grad_outputs):
         t, gloss = inputs[1], grad_outputs[0]
+        n_unit = int(numpy.prod(self.y.shape[2:]))
         gx = cuda.empty_like(self.y)
-        coeff = gloss / t.size
+        if getattr(self, "normalize", True):
```
Comment (beam2d) [inline chainer/functions/softmax_cross_entropy.py]: ditto

File: chainer/functions/softmax_cross_entropy.py [hunk 9]
```diff
@@ -38,49 +39,104 @@ def check_type_backward(self, in_types, out_types):
...
                float* gx, const float* y, const int* t, const float* coeff,
-               int n_channel
+               int n_channel, int n_unit
             ''',
-            'gx[i] = *coeff * (y[i] - ((i % n_channel) == t[i / n_channel]))',
-            'softmax_crossent_bwd')(gx, self.y, t, coeff, self.y.shape[1])
+            '''
+               const int n = i / (n_channel * n_unit);
+               const int c = (i % (n_channel * n_unit)) / n_unit;
+               const int m = (i % (n_channel * n_unit)) % n_unit;
```
Comment (beam2d) [inline chainer/functions/softmax_cross_entropy.py]: ``m = i % n_unit`` is enough

File: tests/functions_tests/test_softmax_cross_entropy.py [hunk 10]
```diff
@@ -33,9 +33,10 @@ def check_forward(self, x_data, t_data, use_cudnn=True):
 
         # Compute expected value
         y = numpy.exp(self.x)
-        loss_expect = 0
+        loss_expect = 0.0
         for i in six.moves.range(y.shape[0]):
-            loss_expect -= math.log(y[i, self.t[i]] / y[i].sum())
+            loss_expect -= math.log(
+                y[i, self.t[i]] / y[i].sum())
```
Comment (beam2d) [inline tests/functions_tests/test_softmax_cross_entropy.py]: Is this line-break needed?

**Comment by beam2d** [comment]
LGTM except above comments! I believe all of them can be fixed quickly.

**Comment by rezoo** [comment]
Thanks for sparing your time for reviewing my PR! I fixed some points you pointed out.

**Comment by beam2d** [comment]
[LGTM](http://lgtm.in/i/SXYZMfh0q)
LGTM!

**Comment by beam2d** [comment]
Thank you for the work! It will be included into v1.2.0.
